kernel-rs/src/vm.rs

```diff
@@ -351,321 +329,300 @@ impl<A: VAddr> PageTable<A> {
-                // TODO(rv6)
-                // We do not know anything about pa, so, for now, we cannot
-                // guarantee that it is safe.
-                kernel().free(unsafe { Page::from_usize(pa) });
-            }
-            pte.invalidate();
+
+        let oldsz = self.size;
+        let mut slf = self;
+        while pgroundup(slf.size) < pgroundup(newsz) {
+            let mut s = scopeguard::guard(slf, |s| {
+                s.dealloc(oldsz);
+            });
+            let mut page = kernel().alloc().ok_or(())?;
+            page.write_bytes(0);
+            s.push_page(page, PteFlags::R | PteFlags::W | PteFlags::X | PteFlags::U)
+                .map_err(|page| kernel().free(page))?;
+            slf = scopeguard::ScopeGuard::into_inner(s);
         }
+        slf.size = newsz;
+        Ok(slf.size)
     }
 
-    /// Deallocate user pages to bring the process size from oldsz to
-    /// newsz.  oldsz and newsz need not be page-aligned, nor does newsz
-    /// need to be less than oldsz.  oldsz can be larger than the actual
-    /// process size.  Returns the new process size.
-    pub fn dealloc(&mut self, oldsz: usize, newsz: usize) -> usize {
-        if newsz >= oldsz {
-            return oldsz;
+    /// Deallocate user pages to bring the process size to newsz, which need
+    /// not be page-aligned. Returns the new process size.
+    pub fn dealloc(&mut self, newsz: usize) -> usize {
+        if self.size <= newsz {
+            return self.size;
         }
 
-        if pgroundup(newsz) < pgroundup(oldsz) {
-            let npages = (pgroundup(oldsz).wrapping_sub(pgroundup(newsz))).wrapping_div(PGSIZE);
-            self.unmap(UVAddr::new(pgroundup(newsz)), npages, true);
+        while pgroundup(newsz) < pgroundup(self.size) {
+            if let Some(page) = self.pop_page() {
+                kernel().free(page);
+            }
         }
+        self.size = newsz;
         newsz
     }
 
-    /// Free user memory pages,
-    /// then free page-table pages.
-    pub fn free(mut self, sz: usize) {
-        if sz > 0 {
-            self.unmap(UVAddr::new(0), pgroundup(sz).wrapping_div(PGSIZE), true);
+    /// Grow or shrink process size by n bytes.
+    /// Return Ok(old size) on success, Err(()) on failure.
+    pub fn resize(&mut self, n: i32) -> Result<usize, ()> {
+        let size = self.size;
+        match n.cmp(&0) {
+            cmp::Ordering::Equal => (),
+            cmp::Ordering::Greater => {
+                self.alloc(size + n as usize)?;
+            }
+            cmp::Ordering::Less => {
+                self.dealloc(size - (-n as usize));
+            }
+        };
+        Ok(size)
+    }
+
+    /// Copy this memory into a given memory. Copies both the page table and
+    /// the physical memory. Returns Ok(()) on success, Err(()) on failure.
+    /// Frees any allocated pages on failure.
+    pub fn copy(&mut self, new: &mut UserMemory) -> Result<(), ()> {
```

Comment:
```suggestion
    pub fn clone(&self) -> Result<Self, ()> {
```
새로운 UserMemory 를 만드는 것이니까 새 것을 직접 리턴하고, self 가 수정되지 않는 것을 보장하면 좋겠습니다.

Reply:
아 근데 `self.get_mut` 을 써야 하네요... ㅠㅠ `&mut self` 로 타협할까요?

Reply:
https://github.com/kaist-cp/rv6/pull/352#discussion_r561998391

Reply:
앗 이미 리뷰한 내용이네요. 그럼 함수 signature 는 그대로 두되 이름만 바꾸면 좋겠습니다. `copy` 는 `copy_in` 과 `copy_out` 때문에 좀 헷갈리기 쉬운 이름인 것 같고, 비슷한 행동을 하는 함수가 standard library 에서는 [`clone_from`](https://doc.rust-lang.org/stable/std/clone/trait.Clone.html#method.clone_from) 이라는 이름을 가지고 있습니다.